File: java/examples/GoodService.java
package com.example.orders.application;

import com.example.orders.domain.Order;
import com.example.orders.domain.OrderStatus;
import com.example.orders.domain.exceptions.OrderNotFoundException;
import com.example.orders.domain.exceptions.OrderNotPayableException;
import com.example.orders.domain.exceptions.PaymentDeclinedException;
import com.example.orders.domain.exceptions.PaymentServiceException;
import com.example.orders.domain.ports.OrderRepository;
import com.example.orders.domain.ports.PaymentGateway;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.Objects;

/**
 * Application service for order payment processing.
 *
 * <p>Orchestrates the payment flow: validates order state, charges the payment
 * gateway, and persists the updated status. All dependencies are injected via
 * constructor — no global state or service locator.
 *
 * <p>This class follows Clean Architecture: it depends only on interfaces
 * defined in the domain layer, not on infrastructure implementations.
 */
@Service
public class OrderPaymentService {

  private static final Logger log = LoggerFactory.getLogger(OrderPaymentService.class);

  private final OrderRepository orderRepository;
  private final PaymentGateway paymentGateway;

  // ✅ Constructor injection — immutable, testable, explicit dependencies
  public OrderPaymentService(OrderRepository orderRepository, PaymentGateway paymentGateway) {
    this.orderRepository = Objects.requireNonNull(orderRepository, "orderRepository must not be null");
    this.paymentGateway = Objects.requireNonNull(paymentGateway, "paymentGateway must not be null");
  }

  /**
   * Process a payment for an existing pending order.
   *
   * <p>Validates the order is in {@link OrderStatus#PENDING} status, charges the
   * payment gateway, and transitions the order to {@link OrderStatus#PAID}.
   *
   * @param orderId   the ID of the order to process; must not be null or blank
   * @param cardToken tokenized card from the payment provider; must not be null or blank
   * @return the payment confirmation ID
   * @throws OrderNotFoundException   if no order with orderId exists
   * @throws OrderNotPayableException if the order is not in PENDING status
   * @throws PaymentDeclinedException if the payment gateway declines the charge
   * @throws PaymentServiceException  if the payment gateway is unreachable
   */
  @Transactional
  public String processPayment(String orderId, String cardToken) {
    // ✅ Guard clauses — fail fast with clear error messages
    validateArguments(orderId, cardToken);

    Order order = orderRepository.findById(orderId)
        .orElseThrow(() -> new OrderNotFoundException(orderId));

    if (order.getStatus() != OrderStatus.PENDING) {
      throw new OrderNotPayableException(orderId, order.getStatus());
    }

    log.info("Processing payment for order {} amount={} currency={}",
        orderId, order.getTotal(), order.getCurrency());

    String paymentId = chargeGateway(orderId, order, cardToken);

    // ✅ Only update status after payment succeeds
    order.markAsPaid(paymentId);
    orderRepository.save(order);

    log.info("Payment processed successfully orderId={} paymentId={}", orderId, paymentId);
    return paymentId;
  }

  // ✅ Private helper — Single Responsibility: charge + translate errors
  private String chargeGateway(String orderId, Order order, String cardToken) {
    try {
      return paymentGateway.charge(order.getTotal(), order.getCurrency(), cardToken);
    } catch (PaymentDeclinedException e) {
      log.info("Payment declined for order {}: {}", orderId, e.getDeclineCode());
      throw e; // re-throw domain exception as-is
    } catch (Exception e) {
      log.error("Payment gateway error for order {}", orderId, e);
      throw new PaymentServiceException("Payment service temporarily unavailable", e);
    }
  }

  // ✅ Validation extracted to named method for readability
  private void validateArguments(String orderId, String cardToken) {
    if (orderId == null || orderId.isBlank()) {
      throw new IllegalArgumentException("orderId must not be null or blank");
    }
    if (cardToken == null || cardToken.isBlank()) {
      throw new IllegalArgumentException("cardToken must not be null or blank");
    }
  }
}

File: kotlin/examples/GoodService.kt
package com.example.orders.application

import com.example.orders.domain.*
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.async
import org.slf4j.LoggerFactory
import java.math.BigDecimal

// ✅ Data class for immutable value objects with validation
data class Money(val amount: BigDecimal, val currency: String) {
    init {
        require(amount >= BigDecimal.ZERO) { "Amount must be non-negative, got: $amount" }
        require(currency.length == 3) { "Currency must be ISO 4217 (3 chars), got: $currency" }
    }

    operator fun plus(other: Money): Money {
        require(currency == other.currency) { "Currency mismatch: $currency vs ${other.currency}" }
        return Money(amount + other.amount, currency)
    }
}

// ✅ Sealed class for exhaustive handling — no else needed in when expressions
sealed class PaymentResult {
    data class Success(val paymentId: String, val amount: BigDecimal) : PaymentResult()
    data class Declined(val code: String, val message: String) : PaymentResult()
    data class ServiceError(val message: String, val cause: Throwable? = null) : PaymentResult()
}

// ✅ Interface for dependency injection — accept interfaces, return concrete types
interface OrderRepository {
    suspend fun findById(orderId: String): Order?
    suspend fun save(order: Order)
}

interface PaymentGateway {
    suspend fun charge(amount: BigDecimal, currency: String, cardToken: String): String
}

// ✅ Service uses constructor injection
class OrderPaymentService(
    private val orderRepository: OrderRepository,
    private val paymentGateway: PaymentGateway,
) {
    private val logger = LoggerFactory.getLogger(OrderPaymentService::class.java)

    /**
     * Process a payment for an existing pending order.
     *
     * @param orderId The ID of the order to process. Must not be blank.
     * @param cardToken Tokenized card from the payment provider.
     * @return [PaymentResult] describing the outcome.
     */
    suspend fun processPayment(orderId: String, cardToken: String): PaymentResult {
        // ✅ require() for precondition validation
        require(orderId.isNotBlank()) { "orderId must not be blank" }
        require(cardToken.isNotBlank()) { "cardToken must not be blank" }

        val order = orderRepository.findById(orderId)
            ?: return PaymentResult.ServiceError("Order $orderId not found")

        if (order.status != OrderStatus.PENDING) {
            return PaymentResult.ServiceError(
                "Order $orderId is not payable (status: ${order.status})"
            )
        }

        logger.info("Processing payment for order $orderId, amount ${order.total}")

        return try {
            val paymentId = paymentGateway.charge(order.total.amount, order.total.currency, cardToken)
            order.status = OrderStatus.PAID
            orderRepository.save(order)

            logger.info("Payment processed: paymentId=$paymentId, orderId=$orderId")
            PaymentResult.Success(paymentId, order.total.amount)
        } catch (e: PaymentDeclinedException) {
            logger.info("Payment declined for order $orderId: ${e.code}")
            PaymentResult.Declined(e.code, e.message ?: "Card declined")
        } catch (e: Exception) {
            logger.error("Payment gateway error for order $orderId", e)
            PaymentResult.ServiceError("Payment service unavailable", e)
        }
    }
}

// ✅ Extension function for concise domain logic
fun PaymentResult.isSuccessful(): Boolean = this is PaymentResult.Success

// ✅ Structured concurrency: parallel data loading
class DashboardService(
    private val orderRepository: OrderRepository,
    private val userService: UserService,
    private val notificationService: NotificationService,
) {
    suspend fun loadDashboard(userId: String): Dashboard = coroutineScope {
        // ✅ async for parallel independent operations
        val ordersDeferred = async { orderRepository.findByUserId(userId) }
        val userDeferred = async { userService.findById(userId) }
        val notificationsDeferred = async { notificationService.getUnread(userId) }

        Dashboard(
            user = userDeferred.await(),
            orders = ordersDeferred.await(),
            notifications = notificationsDeferred.await(),
        )
    }
}
